### PR TITLE
Issue #70 - Add Serverless Support

### LINKED ISSUE #70: [FEATURE] SQL Serverless support
**Is your feature request related to a problem?**
When using the ODBC driver to connect to a Serverless SQL instance, the driver fails to connect. This is because Serverless has a [restricted API set](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/serverless-genref.html), including removing several of the status APIs that the driver uses to test for API health. We already have example implementations for both [SQL CLI](https://github.com/opensearch-project/sql-cli/pull/29) and [SQL JDBC](https://github.com/opensearch-project/sql-jdbc/pull/125) that skip these checks in a serverless environment.

**What solution would you like?**
When a serverless link is provided instead of a standard cluster link, the driver should skip calling status endpoints on serverless so it can connect correctly.

**What alternatives have you considered?**
N/A

**Do you have any additional context?**
N/A

### PR DESCRIPTION
**Pull request by currantw**
### Description

Adds support for OpenSearch Serverless to the ODBC driver.
 
### Issues Resolved

#70 
 
### Check List

- [x] New functionality includes testing.
_Automated testing has been deferred to future issue (see below). See comments for summary of manual testing._
- [x] All tests pass, including unit test, integration test and doctest.
_Pipeline job `build-mac` is failing. This is an existing failure on `main`, and I have raised a new issue to fix it (see below)_
- [x] New functionality has been documented.
- [x] New functionality has javadoc added - _Not applicable_ 
- [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### Related Issues

- #74 
- #75
- #76

### CODE REVIEW

File: src/sqlodbc/dlg_specific.c [hunk 1]
```diff
@@ -37,20 +37,43 @@ void makeConnectString(char *connect_string, const ConnInfo *ci, UWORD len) {
...
+        INI_AUTH_MODE "=%s;"
+        INI_REGION "=%s;"
+        INI_TUNNEL_HOST "=%s;"
+        INI_IS_SERVERLESS "=%d"
+        INI_SSL_USE "=%d;"
+        INI_SSL_HOST_VERIFY "=%d;"
+        INI_LOG_LEVEL "=%d;"
+        INI_LOG_OUTPUT "=%s;"
+        INI_TIMEOUT "=%s;"
+        INI_FETCH_SIZE "=%s;";
```
Comment (currantw) [inline src/sqlodbc/dlg_specific.c:57]: Just pulled this out to try to make it a bit more readable.

File: src/sqlodbc/opensearch_communication.cpp [hunk 2]
```diff
@@ -1048,15 +1071,38 @@ std::string OpenSearchCommunication::GetClusterName() {
...
     }
 }
+
+/**
+ * Returns whether this is connecting to an OpenSearch Serverless cluster.
+ */
+bool OpenSearchCommunication::isServerless() {
+
+    // Specified in DSN configuration.
+    const std::string& is_serverless = m_rt_opts.conn.is_serverless;
```
Comment (jduo) [inline src/sqlodbc/opensearch_communication.cpp:1100]: Can we make the is_serverless field itself a boolean and evaluate it earlier? Rather than every time isServerless() is called? It looks like isServerless() is called during IssueRequest() which is used repeatedly (for executing queries and paging through results).

File: src/sqlodbc/opensearch_communication.h [hunk 3]
```diff
@@ -81,6 +81,7 @@ class OpenSearchCommunication {
     void SetErrorDetails(std::string reason, std::string message,
                          ConnErrorType error_type);
     void SetErrorDetails(ErrorDetails details);
+    bool isServerless();
```
Comment (jduo) [inline src/sqlodbc/opensearch_communication.h:84]: Probably should be marked const since it shouldn't mutate state. Capital 'i' as well.

File: src/sqlodbc/opensearch_communication.cpp (hunk 2 above)
Comment (currantw) [reply src/sqlodbc/opensearch_communication.cpp]: Good call. Added `is_serverless` member variable and a corresponding `SetIsServerless` method, which is called during `EstablishConnection` - so only once.

File: src/sqlodbc/opensearch_communication.h (hunk 3 above)
Comment (currantw) [reply src/sqlodbc/opensearch_communication.h]: Overtaken by subsequent changes. Has been replaced with non-`const` `SetIsServerless` method.

**Comment by currantw** [comment]
Note: The `build_mac` action is also failing on the main project (https://github.com/opensearch-project/sql-odbc). Currently, building locally on Mac fails unless the `-Werror` flag is removed, so this is perhaps the cause of this failure. I will be raising an issue to fix this.

File: src/sqlodbc/opensearch_communication.cpp [hunk 4]
```diff
@@ -548,26 +564,31 @@ bool OpenSearchCommunication::CheckSQLPluginAvailability() {
...
         InitializeConnection();
     }
 
-    // check if the endpoint is initialized
-    if (sql_endpoint.empty()) {
-        SetSqlEndpoint();
+    SetIsServerless();
+    SetSqlEndpoint();
+
+    if (sql_endpoint == SQL_ENDPOINT_ERROR) {
```
Comment (jduo) [inline src/sqlodbc/opensearch_communication.cpp:578]: Can we add a comment here clarifying how connecting to serverless affects these checks? It's hard to follow.

File: docs/user/configuration_options.md [hunk 5]
```diff
@@ -27,9 +27,10 @@
...
-| `HostnameVerification` | Indicate whether certificate hostname verification should be performed for an SSL/TLS connection.                                        | boolean | `0` or `1`                      | `1` (true)  |
+| `HostnameVerification` | Whether certificate hostname verification should be performed for an SSL/TLS connection.                                                 | boolean | `0` or `1`                      | `1` (true)  |
 | `ResponseTimeout`      | The maximum time to wait for responses from the `Host`, in seconds.                                                                      | integer |                                 | `10`        |
 | `FetchSize`            | The page size for all cursor requests. The default value (-1) uses server-defined page size. Set FetchSize to 0 for non-cursor behavior. | integer | `-1`, `0` or any positive value | `-1`        |
+| `IsServerless`         | Whether the connection is to an OpenSearch Serverless cluster. If not specified, it is determined by parsing the server URL.             | boolean | `0` or `1`                      |             |
```
[the remark this answers is not in the document]
Comment (currantw) [reply docs/user/configuration_options.md]: Renamed to `ServerlessOverride`.

File: src/sqlodbc/opensearch_communication.cpp (hunk 4 above)
Comment (currantw) [reply src/sqlodbc/opensearch_communication.cpp]: Added some comments. Also update this `if` clause to also check `is_serverless` (even though this is unnecessary, it hopefully makes the dependencies a bit more clear). Let me know what you think!

**Comment by currantw** [comment]
Automated testing is non-trivial, and has been deferred to #75. In the meantime, we have performed the following to test this new behaviour on Mac:

1. Created a new OpenSearch Serverless cluster that includes SQL.
2. Updated my AWS credentials file ( to add `opensearchodbc` profile with the appropriate access key, secret access key, and session token.
3. Load some test data to OpenSearch using Postman.
4. Downloaded the repo and build the `.dll` using CMake.
5. Updated my `odbc.ini` and `odbcinst.ini` files to configure the ODBC (see below)
6. Download [iodbctest](https://linux.die.net/man/1/iodbctest) (`brew install libiodbc`).
7. Run `iodbctest`, connect to the corresponding DSN, and query the test data.

image

```
# odbc.ini

[ODBC Data Sources]
OpenSearchServerlessDSN = OpenSearchODBC

[OpenSearchServerlessDSN]
Host                 = https://<HOST-URL>.aoss.amazonaws.com
Port                 = 443
UseSSL               = 1
ResponseTimeout      = 10
HostnameVerification = 1
Region               = <AWS_REGION>
Auth                 = AWS_SIGV4
LogOutput            = /Library/ODBC/opensearch-sql-odbc
Driver               = <PROJECT_ROOT>/build/odbc/lib/libsqlodbc.dylib
ServerlessOverride   = 1

# odbcinst.ini
[ODBC Drivers]
OpenSearchODBC = Installed

[OpenSearchODBC]
Driver = <PROJECT_ROOT>/build/odbc/lib/libsqlodbc.dylib
Setup  = <PROJECT_ROOT>//build/odbc/lib/libsqlodbc.dylib
```

File: src/sqlodbc/opensearch_communication.cpp [hunk 6]
```diff
@@ -1046,17 +1075,41 @@ std::string OpenSearchCommunication::GetClusterName() {
...
+        return;
+    }
+
     std::string distribution = GetServerDistribution();
     if (distribution.empty()) {
-        sql_endpoint = SQL_ENDPOINT_ERROR_STR;
-    } else if (distribution.compare("opensearch") == 0) {
-        sql_endpoint = "/_plugins/_sql";
+        sql_endpoint = SQL_ENDPOINT_ERROR;
+    } else if (distribution == "opensearch") {
```
Comment (acarbonetto) [inline src/sqlodbc/opensearch_communication.cpp:1093]: defined "opensearch" as a static constant on the top of the file

File: src/sqlodbc/opensearch_communication.h [hunk 7]
```diff
@@ -82,13 +81,17 @@ class OpenSearchCommunication {
...
+    void SetSqlEndpoint();
+
     // TODO #35 - Go through and add error messages on exit conditions
     std::string m_error_message;
     const std::vector< std::string > m_supported_client_encodings = {"UTF8"};
 
     ConnStatusType m_status;
     ConnErrorType m_error_type;
     std::shared_ptr< ErrorDetails > m_error_details;
+    bool is_serverless;
```
Comment (acarbonetto) [inline src/sqlodbc/opensearch_communication.h:94]: why not is_aoss_serverless?

File: src/sqlodbc/opensearch_communication.cpp (hunk 6 above)
Comment (currantw) [reply src/sqlodbc/opensearch_communication.cpp]: Thanks. Done ✅ 

Aside: Would probably be more robust to also add a `DISTRIBUTION_ELASTICSEARCH` constant as well, and then restructure to below. But I don't know what the constant would be for ElasticSearch, and don't want to get side-tracked by this.

```
std::string distribution = GetServerDistribution();
  if (distribution == DISTRIBUTION_OPENSEARCH) {
    sql_endpoint = SQL_ENDPOINT_OPENSEARCH;
  } else if (distribution == DISTRIBUTION_ELASTICSEARCH) {
    sql_endpoint = SQL_ENDPOINT_ELASTICSEARCH;
  } else {
    sql_endpoint = SQL_ENDPOINT_ERROR;
  }
```

File: src/sqlodbc/opensearch_communication.h (hunk 7 above)
Comment (currantw) [reply src/sqlodbc/opensearch_communication.h:94]: I don't feel strongly about it. Serverless only exists as part of Amazon OpenSearch Service, so the "aoss" doesn't necessary add anything, but not a bad idea to be more explicit/tie it more clearly to the `aoss.amazonaws.com` URL component. 🤷

**Review by MaxKsyunz — CHANGES_REQUESTED** [review]

File: src/sqlodbc/opensearch_communication.h (hunk 7 above)
Comment (acarbonetto) [reply src/sqlodbc/opensearch_communication.h:94]: Always be explicit where possible.

Comment (currantw) [reply src/sqlodbc/opensearch_communication.h]: Updated to `is_aoss_serverless`. Also updated corresponding variable, constant, and function names for consistency.

File: src/sqlodbc/opensearch_types.h [hunk 8]
```diff
@@ -272,6 +272,7 @@ typedef struct connection_options {
     std::string port;
     std::string timeout;
     std::string fetch_size;
+    std::string is_serverless;
```
[the remark this answers is not in the document]
Comment (currantw) [reply src/sqlodbc/opensearch_types.h]: Done! ✅

**Comment by currantw** [comment]
> Automated testing is non-trivial, and has been deferred to #75. In the meantime, we have performed the following to test this new behaviour on Mac:
> 
> **TESTING SERVERLESS**
> 
> Amazon OpenSearch Serverless cluster was setup for testing. See comments for more details.
> 
> ✅ ODBC connects successfully with `iodbctest`.
> 
> **TESTING NON-SERVERLESS WITH EDITED HOST FILE**
> 
> A local OpenSearch cluster was setup for testing.
> 
> ✅ ODBC connects successfully with `iodbctest`.
> 
> **NOTES ON TESTING**
> 
> The following steps where taken to setup and test the Serverless cluster:
> 
> 1. Created a new OpenSearch Serverless cluster that includes SQL.
> 2. Updated my AWS credentials file ( to add `opensearchodbc` profile with the appropriate access key, secret access key, and session token.
> 3. Load some test data to OpenSearch using Postman.
> 4. Downloaded the repo and build the `.dll` using CMake.
> 5. Updated my `odbc.ini` and `odbcinst.ini` files to configure the ODBC (see below)
> 6. Download [iodbctest](https://linux.die.net/man/1/iodbctest) (`brew install libiodbc`).
> 7. Run `iodbctest`, connect to the corresponding DSN, and query the test data.
> 
> image
> ```
> # odbc.ini
> 
> [ODBC Data Sources]
> OpenSearchServerlessDSN = OpenSearchODBC
> 
> [OpenSearchServerlessDSN]
> Host                 = https://<HOST-URL>.aoss.amazonaws.com
> Port                 = 443
> UseSSL               = 1
> ResponseTimeout      = 10
> HostnameVerification = 1
> Region               = <AWS_REGION>
> Auth                 = AWS_SIGV4
> LogOutput            = /Library/ODBC/opensearch-sql-odbc
> Driver               = <PROJECT_ROOT>/build/odbc/lib/libsqlodbc.dylib
> ServerlessOverride   = 1
> 
> # odbcinst.ini
> [ODBC Drivers]
> OpenSearchODBC = Installed
> 
> [OpenSearchODBC]
> Driver = <PROJECT_ROOT>/build/odbc/lib/libsqlodbc.dylib
> Setup  = <PROJECT_ROOT>//build/odbc/lib/libsqlodbc.dylib
> ```

**Review by MaxKsyunz — APPROVED** [review]
